File: ios/RNGoogleMapsPlusView.swift
import CoreLocation
import Foundation
import GoogleMaps
import NitroModules

final class RNGoogleMapsPlusView: HybridRNGoogleMapsPlusViewSpec {

  private let mapErrorHandler: MapErrorHandler
  private let permissionHandler: PermissionHandler
  private let locationHandler: LocationHandler

  private let impl: GoogleMapsViewImpl

  var view: UIView {
    return impl
  }

  override init() {
    self.permissionHandler = PermissionHandler()
    self.locationHandler = LocationHandler()
    self.mapErrorHandler = MapErrorHandler()
    self.impl = GoogleMapsViewImpl(
      mapErrorHandler: mapErrorHandler,
      locationHandler: locationHandler
    )
  }

  func onDropView() {
    impl.deinitInternal()
  }

  var initialProps: RNInitialProps? {
    didSet {
      let options = GMSMapViewOptions()
      initialProps?.mapId.map { options.mapID = GMSMapID(identifier: $0) }
      initialProps?.liteMode.map { _ in /* not supported */ }
      initialProps?.camera.map {
        options.camera = $0.toGMSCameraPosition(current: nil)
      }
      initialProps?.backgroundColor.map {
        options.backgroundColor = $0.toUIColor()
      }

      impl.googleMapOptions = options
    }
  }

  var uiSettings: RNMapUiSettings? {
    didSet { impl.uiSettings = uiSettings }
  }

  var myLocationEnabled: Bool? {
    didSet { impl.myLocationEnabled = myLocationEnabled }
  }

  var buildingEnabled: Bool? {
    didSet { impl.buildingEnabled = buildingEnabled }
  }

  var trafficEnabled: Bool? {
    didSet { impl.trafficEnabled = trafficEnabled }
  }

  var indoorEnabled: Bool? {
    didSet { impl.indoorEnabled = indoorEnabled }
  }

  var transitEnabled: Bool? {
    didSet { impl.transitEnabled = transitEnabled }
  }

  var customMapStyle: String? {
    didSet {
      if let value = customMapStyle {
        impl.customMapStyle = try? GMSMapStyle(jsonString: value)
      }
    }
  }

  var userInterfaceStyle: RNUserInterfaceStyle? {
    didSet {
      impl.userInterfaceStyle = userInterfaceStyle?.toUIUserInterfaceStyle
    }
  }

  var mapZoomConfig: RNMapZoomConfig? {
    didSet { impl.mapZoomConfig = mapZoomConfig }
  }

  var mapPadding: RNMapPadding? {
    didSet { impl.mapPadding = mapPadding }
  }

  var mapType: RNMapType? {
    didSet {
      impl.mapType = mapType?.toGMSMapViewType
    }
  }

  var markers: [RNMarker]? {
    didSet {
      let prevById = Dictionary(
        (oldValue ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )
      let nextById = Dictionary(
        (markers ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )

      let removed = Set(prevById.keys).subtracting(nextById.keys)
      removed.forEach { impl.removeMarker(id: $0) }

      for (id, next) in nextById {
        if prevById[id] == nil {
          impl.addMarker(next)
        } else {
          impl.updateMarker(next)
        }
      }
    }
  }

  var polylines: [RNPolyline]? {
    didSet {
      let prevById = Dictionary(
        (oldValue ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )
      let nextById = Dictionary(
        (polylines ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )

      let removed = Set(prevById.keys).subtracting(nextById.keys)
      removed.forEach { impl.removePolyline(id: $0) }

      for (id, next) in nextById {
        if prevById[id] == nil {
          impl.addPolyline(next)
        } else {
          impl.updatePolyline(next)
        }
      }
    }
  }

  var polygons: [RNPolygon]? {
    didSet {
      let prevById = Dictionary(
        (oldValue ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )
      let nextById = Dictionary(
        (polygons ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )

      let removed = Set(prevById.keys).subtracting(nextById.keys)
      removed.forEach { impl.removePolygon(id: $0) }

      for (id, next) in nextById {
        if prevById[id] == nil {
          impl.addPolygon(next)
        } else {
          impl.updatePolygon(next)
        }
      }
    }
  }

  var circles: [RNCircle]? {
    didSet {
      let prevById = Dictionary(
        (oldValue ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )
      let nextById = Dictionary(
        (circles ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )

      let removed = Set(prevById.keys).subtracting(nextById.keys)
      removed.forEach { impl.removeCircle(id: $0) }

      for (id, next) in nextById {
        if prevById[id] == nil {
          impl.addCircle(next)
        } else {
          impl.updateCircle(next)
        }
      }
    }
  }

  var heatmaps: [RNHeatmap]? {
    didSet {
      let prevById = Dictionary(
        (oldValue ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )
      let nextById = Dictionary(
        (heatmaps ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )

      let removed = Set(prevById.keys).subtracting(nextById.keys)
      removed.forEach { impl.removeHeatmap(id: $0) }

      for (id, next) in nextById {
        if prevById[id] == nil {
          impl.addHeatmap(next)
        } else {
          impl.updateHeatmap(next)
        }
      }
    }
  }

  var kmlLayers: [RNKMLayer]? {
    didSet {
      let prevById = Dictionary(
        (oldValue ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )
      let nextById = Dictionary(
        (kmlLayers ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )

      let removed = Set(prevById.keys).subtracting(nextById.keys)
      removed.forEach { impl.removeKmlLayer(id: $0) }

      for (id, next) in nextById {
        if prevById[id] == nil {
          impl.addKmlLayer(next)
        } else {
          impl.updateKmlLayer(next)
        }
      }
    }
  }

  var urlTileOverlays: [RNUrlTileOverlay]? {
    didSet {
      let prevById = Dictionary(
        (oldValue ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )
      let nextById = Dictionary(
        (urlTileOverlays ?? []).map { ($0.id, $0) },
        uniquingKeysWith: { _, new in new }
      )

      let removed = Set(prevById.keys).subtracting(nextById.keys)
      removed.forEach { impl.removeUrlTileOverlay(id: $0) }

      for (id, next) in nextById {
        if prevById[id] == nil {
          impl.addUrlTileOverlay(next)
        } else {
          impl.updateUrlTileOverlay(next)
        }
      }
    }
  }

  var locationConfig: RNLocationConfig? {
    didSet {
      impl.locationConfig = locationConfig
    }
  }

  var onMapError: ((RNMapErrorCode, String) -> Void)? {
    didSet { mapErrorHandler.callback = onMapError }
  }
  var onMapReady: ((Bool) -> Void)? {
    didSet { impl.onMapReady = onMapReady }
  }
  var onMapLoaded: ((RNRegion, RNCamera) -> Void)? {
    didSet { impl.onMapLoaded = onMapLoaded }
  }
  var onLocationUpdate: ((RNLocation) -> Void)? {
    didSet { impl.onLocationUpdate = onLocationUpdate }
  }
  var onLocationError: ((_ error: RNLocationErrorCode) -> Void)? {
    didSet { impl.onLocationError = onLocationError }
  }
  var onMapPress: ((RNLatLng) -> Void)? {
    didSet { impl.onMapPress = onMapPress }
  }
  var onMapLongPress: ((RNLatLng) -> Void)? {
    didSet { impl.onMapLongPress = onMapLongPress }
  }
  var onPoiPress: ((String, String, RNLatLng) -> Void)? {
    didSet { impl.onPoiPress = onPoiPress }
  }
  var onMarkerPress: ((String) -> Void)? {
    didSet { impl.onMarkerPress = onMarkerPress }
  }
  var onPolylinePress: ((String) -> Void)? {
    didSet { impl.onPolylinePress = onPolylinePress }
  }
  var onPolygonPress: ((String) -> Void)? {
    didSet { impl.onPolygonPress = onPolygonPress }
  }
  var onCirclePress: ((String) -> Void)? {
    didSet { impl.onCirclePress = onCirclePress }
  }
  var onMarkerDragStart: ((String, RNLatLng) -> Void)? {
    didSet { impl.onMarkerDragStart = onMarkerDragStart }
  }
  var onMarkerDrag: ((String, RNLatLng) -> Void)? {
    didSet { impl.onMarkerDrag = onMarkerDrag }
  }
  var onMarkerDragEnd: ((String, RNLatLng) -> Void)? {
    didSet { impl.onMarkerDragEnd = onMarkerDragEnd }
  }
  var onIndoorBuildingFocused: ((RNIndoorBuilding) -> Void)? {
    didSet { impl.onIndoorBuildingFocused = onIndoorBuildingFocused }
  }
  var onIndoorLevelActivated: ((RNIndoorLevel) -> Void)? {
    didSet { impl.onIndoorLevelActivated = onIndoorLevelActivated }
  }
  var onInfoWindowPress: ((String) -> Void)? {
    didSet { impl.onInfoWindowPress = onInfoWindowPress }
  }
  var onInfoWindowClose: ((String) -> Void)? {
    didSet { impl.onInfoWindowClose = onInfoWindowClose }
  }
  var onInfoWindowLongPress: ((String) -> Void)? {
    didSet { impl.onInfoWindowLongPress = onInfoWindowLongPress }
  }
  var onMyLocationPress: ((RNLocation) -> Void)? {
    didSet { impl.onMyLocationPress = onMyLocationPress }
  }
  var onMyLocationButtonPress: ((Bool) -> Void)? {
    didSet { impl.onMyLocationButtonPress = onMyLocationButtonPress }
  }
  var onCameraChangeStart: ((RNRegion, RNCamera, Bool) -> Void)? {
    didSet { impl.onCameraChangeStart = onCameraChangeStart }
  }
  var onCameraChange: ((RNRegion, RNCamera, Bool) -> Void)? {
    didSet { impl.onCameraChange = onCameraChange }
  }
  var onCameraChangeComplete: ((RNRegion, RNCamera, Bool) -> Void)? {
    didSet { impl.onCameraChangeComplete = onCameraChangeComplete }
  }

  func showMarkerInfoWindow(id: String) {
    impl.showMarkerInfoWindow(id: id)
  }

  func hideMarkerInfoWindow(id: String) {
    impl.hideMarkerInfoWindow(id: id)
  }

  func clearMarkerIconCache() {
    impl.clearMarkerIconCache()
  }

  func setCamera(camera: RNCameraUpdate, animated: Bool?, durationMs: Double?) {
    onMain {
      let cam = camera.toGMSCameraPosition(current: self.impl.currentCamera)
      self.impl.setCamera(
        camera: cam,
        animated: animated ?? false,
        durationMs: durationMs ?? 3000
      )
    }
  }

  func setCameraToCoordinates(
    coordinates: [RNLatLng],
    padding: RNMapPadding?,
    animated: Bool?,
    durationMs: Double?
  ) {
    impl.setCameraToCoordinates(
      coordinates: coordinates,
      padding: padding ?? RNMapPadding(0, 0, 0, 0),
      animated: animated ?? false,
      durationMs: durationMs ?? 3000
    )
  }

  func setCameraBounds(bounds: RNLatLngBounds?) {
    impl.setCameraBounds(bounds?.toCoordinateBounds())
  }

  func animateToBounds(
    bounds: RNLatLngBounds,
    padding: Double?,
    durationMs: Double?,
    lockBounds: Bool?
  ) {
    impl.animateToBounds(
      bounds.toCoordinateBounds(),
      padding: padding ?? 0,
      durationMs: durationMs ?? 3000,
      lockBounds: lockBounds ?? false
    )
  }

  func snapshot(
    options: RNSnapshotOptions,
  ) -> NitroModules.Promise<String?> {
    return impl.snapshot(
      size: options.size?.toCGSize(),
      format: options.format.toFileExtension(),
      imageFormat: options.format.toImageFormat(),
      quality: CGFloat(options.quality),
      resultIsFile: options.resultType.isFileResult()
    )

  }

  func showLocationDialog() {
    locationHandler.showLocationDialog()
  }

  func openLocationSettings() {
    locationHandler.openLocationSettings()
  }

  func requestLocationPermission()
  -> NitroModules.Promise<RNLocationPermissionResult> {
    return permissionHandler.requestLocationPermission()
  }

  func isGooglePlayServicesAvailable() -> Bool {
    /// not supported
    return false
  }
}

File: android/src/main/java/com/rngooglemapsplus/RNGoogleMapsPlusView.kt
package com.rngooglemapsplus

import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import com.google.android.gms.maps.GoogleMapOptions
import com.google.android.gms.maps.model.MapStyleOptions
import com.margelo.nitro.core.Promise
import com.rngooglemapsplus.extensions.isFileResult
import com.rngooglemapsplus.extensions.toCameraPosition
import com.rngooglemapsplus.extensions.toColor
import com.rngooglemapsplus.extensions.toCompressFormat
import com.rngooglemapsplus.extensions.toFileExtension
import com.rngooglemapsplus.extensions.toGoogleMapType
import com.rngooglemapsplus.extensions.toLatLngBounds
import com.rngooglemapsplus.extensions.toMapColorScheme
import com.rngooglemapsplus.extensions.toMarkerTag
import com.rngooglemapsplus.extensions.toSize

@DoNotStrip
class RNGoogleMapsPlusView(
  private val context: ThemedReactContext,
) : HybridRNGoogleMapsPlusViewSpec() {
  private val mapErrorHandler = MapErrorHandler()

  private val permissionHandler = PermissionHandler(context)
  private val locationHandler = LocationHandler(context)
  private val playServiceHandler = PlayServicesHandler(context)

  override val view =
    GoogleMapsViewImpl(context, locationHandler, playServiceHandler, mapErrorHandler)

  override fun onDropView() {
    view.destroyInternal()
  }

  override var initialProps: RNInitialProps? = null
    set(value) {
      if (field == value) return
      field = value

      val options =
        GoogleMapOptions().apply {
          initialProps?.mapId?.let { mapId(it) }
          initialProps?.liteMode?.let { liteMode(it) }
          initialProps?.camera?.let { camera(it.toCameraPosition(current = null)) }
          initialProps?.backgroundColor?.let { backgroundColor(it.toColor()) }
        }

      view.googleMapsOptions = options
    }

  override var uiSettings: RNMapUiSettings? = null
    set(value) {
      if (field == value) return
      field = value
      view.uiSettings = value
    }

  override var myLocationEnabled: Boolean? = null
    set(value) {
      if (field == value) return
      field = value
      view.myLocationEnabled = value
    }

  override var buildingEnabled: Boolean? = null
    set(value) {
      if (field == value) return
      field = value
      view.buildingEnabled = value
    }

  override var trafficEnabled: Boolean? = null
    set(value) {
      if (field == value) return
      field = value
      view.trafficEnabled = value
    }

  override var indoorEnabled: Boolean? = null
    set(value) {
      if (field == value) return
      field = value
      view.indoorEnabled = value
    }

  override var transitEnabled: Boolean? = null
    set(value) {
      if (field == value) return
      field = value
      view.transitEnabled = value
    }

  override var customMapStyle: String? = null
    set(value) {
      if (field == value) return
      field = value
      value?.let {
        view.customMapStyle = MapStyleOptions(it)
      }
    }

  override var userInterfaceStyle: RNUserInterfaceStyle? = null
    set(value) {
      if (field == value) return
      field = value
      view.userInterfaceStyle = value.toMapColorScheme()
    }

  override var mapZoomConfig: RNMapZoomConfig? = null
    set(value) {
      if (field == value) return
      field = value
      view.mapZoomConfig = value
    }

  override var mapPadding: RNMapPadding? = null
    set(value) {
      if (field == value) return
      field = value
      view.mapPadding = value
    }

  override var mapType: RNMapType? = null
    set(value) {
      if (field == value) return
      field = value
      view.mapType = value?.toGoogleMapType()
    }

  override var markers: Array<RNMarker>? = null
    set(value) {
      if (field.contentEquals(value)) return
      val prevById = field?.associateBy { it.id } ?: emptyMap()
      val nextById = value?.associateBy { it.id } ?: emptyMap()
      field = value

      (prevById.keys - nextById.keys).forEach { id ->
        view.removeMarker(id)
      }

      nextById.forEach { (id, next) ->
        val prev = prevById[id]
        when {
          prev == null -> {
            markerBuilder.buildIconAsync(next) { icon, hitbox ->
              view.addMarker(
                id,
                markerBuilder.build(next, icon),
                next.toMarkerTag(hitbox),
              )
            }
          }

          !prev.markerEquals(next) -> {
            if (markerBuilder.hasIconJob(id)) {
              markerBuilder.buildIconAsync(next) { icon, hitbox ->
                view.addMarker(
                  id,
                  markerBuilder.build(next, icon),
                  next.toMarkerTag(hitbox),
                )
              }
            } else {
              view.updateMarker(id) { marker ->
                markerBuilder.update(prev, next, marker)
              }
            }
          }
        }
      }
    }

  override var polylines: Array<RNPolyline>? = null
    set(value) {
      if (field.contentEquals(value)) return
      val prevById = field?.associateBy { it.id } ?: emptyMap()
      val nextById = value?.associateBy { it.id } ?: emptyMap()
      field = value
      (prevById.keys - nextById.keys).forEach { id ->
        view.removePolyline(id)
      }

      nextById.forEach { (id, next) ->
        if (prevById[id] == null) {
          view.addPolyline(next)
        } else {
          view.updatePolyline(next)
        }
      }
    }

  override var polygons: Array<RNPolygon>? = null
    set(value) {
      if (field.contentEquals(value)) return
      val prevById = field?.associateBy { it.id } ?: emptyMap()
      val nextById = value?.associateBy { it.id } ?: emptyMap()
      field = value

      (prevById.keys - nextById.keys).forEach { id ->
        view.removePolygon(id)
      }

      nextById.forEach { (id, next) ->
        if (prevById[id] == null) {
          view.addPolygon(next)
        } else {
          view.updatePolygon(next)
        }
      }
    }

  override var circles: Array<RNCircle>? = null
    set(value) {
      if (field.contentEquals(value)) return
      val prevById = field?.associateBy { it.id } ?: emptyMap()
      val nextById = value?.associateBy { it.id } ?: emptyMap()
      field = value

      (prevById.keys - nextById.keys).forEach { id ->
        view.removeCircle(id)
      }

      nextById.forEach { (id, next) ->
        if (prevById[id] == null) {
          view.addCircle(next)
        } else {
          view.updateCircle(next)
        }
      }
    }

  override var heatmaps: Array<RNHeatmap>? = null
    set(value) {
      if (field.contentEquals(value)) return
      val prevById = field?.associateBy { it.id } ?: emptyMap()
      val nextById = value?.associateBy { it.id } ?: emptyMap()
      field = value
      (prevById.keys - nextById.keys).forEach { id ->
        view.removeHeatmap(id)
      }

      nextById.forEach { (id, next) ->
        if (prevById[id] == null) {
          view.addHeatmap(next)
        } else {
          view.updateHeatmap(next)
        }
      }
    }

  override var kmlLayers: Array<RNKMLayer>? = null
    set(value) {
      if (field.contentEquals(value)) return
      val prevById = field?.associateBy { it.id } ?: emptyMap()
      val nextById = value?.associateBy { it.id } ?: emptyMap()
      field = value
      (prevById.keys - nextById.keys).forEach { id ->
        view.removeKmlLayer(id)
      }
      nextById.forEach { (id, next) ->
        if (prevById[id] == null) {
          view.addKmlLayer(next)
        } else {
          view.updateKmlLayer(next)
        }
      }
    }

  override var urlTileOverlays: Array<RNUrlTileOverlay>? = null
    set(value) {
      if (field.contentEquals(value)) return
      val prevById = field?.associateBy { it.id } ?: emptyMap()
      val nextById = value?.associateBy { it.id } ?: emptyMap()
      field = value
      (prevById.keys - nextById.keys).forEach { id ->
        view.removeUrlTileOverlay(id)
      }

      nextById.forEach { (id, next) ->
        if (prevById[id] == null) {
          view.addUrlTileOverlay(next)
        } else {
          view.updateUrlTileOverlay(next)
        }
      }
    }

  override var locationConfig: RNLocationConfig? = null
    set(value) {
      if (field == value) return
      field = value
      view.locationConfig = value
    }

  override var onMapError: ((RNMapErrorCode, String) -> Unit)? = null
    set(cb) {
      field = cb
      mapErrorHandler.callback = cb
    }

  override var onMapReady: ((Boolean) -> Unit)? = null
    set(cb) {
      view.onMapReady = cb
    }

  override var onMapLoaded: ((RNRegion, RNCamera) -> Unit)? = null
    set(cb) {
      view.onMapLoaded = cb
    }

  override var onLocationUpdate: ((RNLocation) -> Unit)? = null
    set(cb) {
      view.onLocationUpdate = cb
    }

  override var onLocationError: ((RNLocationErrorCode) -> Unit)? = null
    set(cb) {
      view.onLocationError = cb
    }

  override var onMapPress: ((RNLatLng) -> Unit)? = null
    set(cb) {
      view.onMapPress = cb
    }

  override var onMapLongPress: ((RNLatLng) -> Unit)? = null
    set(cb) {
      view.onMapLongPress = cb
    }

  override var onMarkerPress: ((String) -> Unit)? = null
    set(cb) {
      view.onMarkerPress = cb
    }

  override var onPoiPress: ((String, String, RNLatLng) -> Unit)? = null
    set(cb) {
      view.onPoiPress = cb
    }

  override var onPolylinePress: ((String) -> Unit)? = null
    set(cb) {
      view.onPolylinePress = cb
    }

  override var onPolygonPress: ((String) -> Unit)? = null
    set(cb) {
      view.onPolygonPress = cb
    }

  override var onCirclePress: ((String) -> Unit)? = null
    set(cb) {
      view.onCirclePress = cb
    }

  override var onMarkerDragStart: ((String, RNLatLng) -> Unit)? = null
    set(cb) {
      view.onMarkerDragStart = cb
    }

  override var onMarkerDrag: ((String, RNLatLng) -> Unit)? = null
    set(cb) {
      view.onMarkerDrag = cb
    }

  override var onMarkerDragEnd: ((String, RNLatLng) -> Unit)? = null
    set(cb) {
      view.onMarkerDragEnd = cb
    }

  override var onIndoorBuildingFocused: ((RNIndoorBuilding) -> Unit)? = null
    set(cb) {
      view.onIndoorBuildingFocused = cb
    }

  override var onIndoorLevelActivated: ((RNIndoorLevel) -> Unit)? = null
    set(cb) {
      view.onIndoorLevelActivated = cb
    }

  override var onInfoWindowPress: ((String) -> Unit)? = null
    set(cb) {
      view.onInfoWindowPress = cb
    }

  override var onInfoWindowClose: ((String) -> Unit)? = null
    set(cb) {
      view.onInfoWindowClose = cb
    }

  override var onInfoWindowLongPress: ((String) -> Unit)? = null
    set(cb) {
      view.onInfoWindowLongPress = cb
    }

  override var onMyLocationPress: ((RNLocation) -> Unit)? = null
    set(cb) {
      view.onMyLocationPress = cb
    }

  override var onMyLocationButtonPress: ((Boolean) -> Unit)? = null
    set(cb) {
      view.onMyLocationButtonPress = cb
    }

  override var onCameraChangeStart: ((RNRegion, RNCamera, Boolean) -> Unit)? = null
    set(cb) {
      view.onCameraChangeStart = cb
    }

  override var onCameraChange: ((RNRegion, RNCamera, Boolean) -> Unit)? = null
    set(cb) {
      view.onCameraChange = cb
    }

  override var onCameraChangeComplete: ((RNRegion, RNCamera, Boolean) -> Unit)? = null
    set(cb) {
      view.onCameraChangeComplete = cb
    }

  override fun showMarkerInfoWindow(id: String) {
    view.showMarkerInfoWindow(id)
  }

  override fun hideMarkerInfoWindow(id: String) {
    view.hideMarkerInfoWindow(id)
  }

  override fun clearMarkerIconCache() {
    view.clearMarkerIconCache()
  }

  override fun setCamera(
    camera: RNCameraUpdate,
    animated: Boolean?,
    durationMs: Double?,
  ) {
    val current = view.currentCamera
    view.setCamera(
      camera.toCameraPosition(current),
      animated ?: false,
      durationMs?.toInt() ?: 3000,
    )
  }

  override fun setCameraToCoordinates(
    coordinates: Array<RNLatLng>,
    padding: RNMapPadding?,
    animated: Boolean?,
    durationMs: Double?,
  ) {
    view.setCameraToCoordinates(
      coordinates,
      padding = padding ?: RNMapPadding(0.0, 0.0, 0.0, 0.0),
      animated ?: false,
      durationMs?.toInt() ?: 3000,
    )
  }

  override fun setCameraBounds(bounds: RNLatLngBounds?) {
    view.setCameraBounds(
      bounds?.toLatLngBounds(),
    )
  }

  override fun animateToBounds(
    bounds: RNLatLngBounds,
    padding: Double?,
    durationMs: Double?,
    lockBounds: Boolean?,
  ) {
    view.animateToBounds(
      bounds.toLatLngBounds(),
      padding = padding?.toInt() ?: 0,
      durationMs?.toInt() ?: 3000,
      lockBounds = lockBounds ?: false,
    )
  }

  override fun snapshot(options: RNSnapshotOptions): Promise<String?> =
    view.snapshot(
      size = options.size.toSize(),
      format = options.format.toFileExtension(),
      compressFormat = options.format.toCompressFormat(),
      quality = options.quality,
      resultIsFile = options.resultType.isFileResult(),
    )

  override fun showLocationDialog() {
    locationHandler.showLocationDialog()
  }

  override fun openLocationSettings() {
    locationHandler.openLocationSettings()
  }

  override fun requestLocationPermission(): Promise<RNLocationPermissionResult> = permissionHandler.requestLocationPermission()

  override fun isGooglePlayServicesAvailable(): Boolean = playServiceHandler.isPlayServicesAvailable()
}
